crypto/cryptlib.c

```diff
@@ -58,25 +63,38 @@ void OPENSSL_cpuid_setup(void)
             }
         } else if (env[0] == ':') {
             vec = OPENSSL_ia32_cpuid(OPENSSL_ia32cap_P);
+        } else {
+#  ifdef OPENSSL_OVERRIDE_CPUID
+            OPENSSL_ia32_cpuid(OPENSSL_ia32cap_P);
+#  else
+            vec = OPENSSL_ia32_cpuid(OPENSSL_ia32cap_P) & vec;
+#  endif
         }
 
         if ((env = strchr(env, ':'))) {
             IA32CAP vecx;
             env++;
             off = (env[0] == '~') ? 1 : 0;
 #  if defined(_WIN32)
-            if (!sscanf(env + off, "%I64i", &vecx))
+            if (sscanf(env + off, "%I64i", &vecx) != 1)
                 vecx = strtoul(env + off, NULL, 0);
 #  else
-            if (!sscanf(env + off, "%lli", (long long *)&vecx))
+            if (sscanf(env + off, "%lli", &llval) != 1)
                 vecx = strtoul(env + off, NULL, 0);
+            else
+                vecx = llval;
 #  endif
             if (off) {
                 OPENSSL_ia32cap_P[2] &= ~(unsigned int)vecx;
```

Comment:
unless I am mistaken OPENSSL_ia32cap=0:~1 did not initialize OPENSSL_ia32cap_P[2..3]
and used an uninitialized value in the &= here

Reply:
That's why I added the OPENSSL_ia32_cpuid call in ifdef-OPENSSL_OVERRIDE_CPUID above.

Reply:
> 0:~1 did not initialize OPENSSL_ia32cap_P[2..3]

Well, it you *set* first quadword, then second one is cleared. *One* of problems is that there are certain relationships between bits in first and second quadwords(*), and instead of tracking them, it's just *easier* to say that if first quadword is set, second one is simply cleared. Yes, you can sense contradiction, because if there are relationships, then masking bits in first quadword with ~ should track these relationships. Admittedly code is a bit heuristic and user should mostly be aware of what's going on. There *are* invalid/inappropriate combinations, but tracking them down is daunting task and it's not clear if it's worth the effort... But it might be just me...

(*) For example if you mask xmm, then you shouldn't be using ymm.